docs/designs/allreduce_based_horovod.md

```diff
@@ -0,0 +1,117 @@
+# Design for Elastic AllReduce-based Training Support Based on Horovod
+
+This document describes the design for supporting AllReduce-based distributed
+training based on Horovod in ElasticDL.
+
+## Motivation
+
+We have developed elastic AllReduce based on FTlib in ElasticDL.
+From the [benchmark report](../benchmark/ftlib_benchmark.md), we can
+find that the performance of FTlib for ResNet50 is worse than Horovod.
+What's more, it is not stable to build consensus by gossip protocol in FTlib
+like the [issue](https://github.com/sql-machine-learning/elasticdl/issues/2192#issuecomment-664096185).
+
+FTlib uses Gloo to implement elastic AllReduce because the worker process
+can catch the exception from Gloo and not exit if the AllReduce operator
+fails. Horovod also can use Gloo as backend. What's more, there are many
+optimizations like [Tensor Fusion](https://horovod.readthedocs.io/en/latest/tensor-fusion_include.html)
+in Horovod to improve performance. So, the performance of Horovod for ResNet50 is
+better than FTlib because ResNet50 has may small tensors.
+
+Horovod provides Python APIs for Gloo and we can use those APIs in ElasticDL
+to implement elastic AllReduce.
+
+## ElasticDL Re-initialize Horovod When the Number of Workers Changes
+
+When using Horovod with Gloo backend, we need to create a `RendezvousServer`
```

Comment:
When using Horovod with Gloo backend, we need to create a `RendezvousServer`, which  has a KVStore. Also we need to put a host plan into this KVStore. A host plan includes worker hosts and their assigned ranks,  which are required by Gloo. 
The master in ElasticDL is responsible to create the `RendezvousServer`.  To support for elastic, when the master detects worker number change, it will create a new host plan and put it in the KVStore.